naptime/src/main/scala/org/coursera/naptime/ari/engine/EngineImpl.scala

```diff
@@ -2,42 +2,108 @@ package org.coursera.naptime.ari.engine
-      allTypes ++ routerBuilder.types.flatMap { keyedSchema =>
-        keyedSchema.value match {
-          case recordSchemaValue: DataSchema => Some(keyedSchema.key -> recordSchemaValue)
-          case _ => None
-        }
-      }
-    }
+  private[this] val mergedTypes = naptimeRoutes.routerBuilders.flatMap(_.types.map(_.tuple))
+    .filter(_._2.isInstanceOf[RecordDataSchema])
+    .map(tuple => tuple._1 -> tuple._2.asInstanceOf[RecordDataSchema]).toMap
 
   override def execute(request: Request): Future[Response] = {
     val responseFutures = request.topLevelRequests.map { topLevelRequest =>
-      val singleRequest = Request(request.requestHeader, List(topLevelRequest))
-      fetcher.data(singleRequest)
+      executeTopLevelRequest(request, topLevelRequest)
     }
     val futureResponses = Future.sequence(responseFutures)
     futureResponses.map { responses =>
       responses.foldLeft(Response.empty)(_ ++ _)
     }
   }
+
+  private[this] def executeTopLevelRequest(request: Request, topLevelRequest: TopLevelRequest): Future[Response] = {
+    val singleRequest = Request(request.requestHeader, List(topLevelRequest))
+    val topLevelResponse = fetcher.data(singleRequest)
+
+    topLevelResponse.flatMap { topLevelResponse =>
+      // If we have a schema, we can perform automatic resource inclusion.
+      (for {
+        topLevelSchema <- resourceSchemas.get(topLevelRequest.resource)
+        mergedType <- mergedTypes.get(topLevelSchema.mergedType)
+      } yield {
+        val topLevelData = topLevelResponse.data.get(topLevelRequest.resource).toIterable.flatMap { dataMap =>
+          dataMap.values
+        }
+        val nestedLookups = topLevelRequest.selection.selections.filter(_.selections.nonEmpty)
+        val additionalResponses = nestedLookups.map { nestedField =>
+          val relatedResource = Option(mergedType.getField(nestedField.name)).flatMap { field =>
+            Option(field.getProperties.get(Relations.PROPERTY_NAME)).map {
+              case idString: String =>
+                ResourceName.parse(idString).getOrElse {
+                  throw new IllegalStateException(s"Could not parse identifier '$idString' for field '$field' in " +
+                    s"merged type $mergedType")
+                }
+              case identifier =>
+                throw new IllegalStateException(s"Unexpected type for identifier '$identifier' for field '$field' " +
+                  s"in merged type $mergedType")
+            }
+          }.getOrElse {
+            throw new IllegalStateException(s"Could not find nested field $nestedField on merged type $mergedType")
+          }
+
+          val multiGetIds = topLevelData.flatMap { elem =>
+            // TODO: support single foreign key relations (currently this supports only nested lists)
+            val field = Option(elem.getDataList(nestedField.name)).map(_.asScala).getOrElse {
```

Comment:
what if we have a one-to-one relationship (a course has a single courseDetails.v1 related thing) -- it looks like you're only supporting datalists here?